### PR TITLE
build(snap): stage ekuiper snap instead of local build

### PR DESCRIPTION
**Pull request by MonicaisHer**
This PR stages a standalone [edgex-ekuiper snap](https://github.com/canonical/edgex-ekuiper-snap) into edgexfoundry snap, and it will replace the exiting kuiper app. This is to ensure backward compatibility for upgrades from 2.0/2.1 to 2.x.

- remove secure message bus setup as config files coming from edgex-ekuiper stage-snap directly
- use configure hook generating redis creds for kuiper
- update kuiper `apps` and `parts`
- update edgex-ekuiper stage-snap track `1/beta` for temporary use (replace it with `1/stable` when possible)

Signed-off-by: Mengyi Wang <mengyi.wang@canonical.com>

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)The [checkbox test ](https://github.com/canonical/edgex-checkbox-provider/blob/master/data/latest/test-rules-engine.sh) has covered the rules engine, and this PR passed it.
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions

1. Install edgex and turn on kuiper service:
```
snapcraft
snap install edgexfoundry_2.0.1-dev.71_amd64.snap --devmode
snap set edgexfoundry device-virtual=on
snap set edgexfoundry kuiper=on
```
2. Create stream and rules to ensure kuiper works:
```
# stream 1
edgexfoundry.kuiper-cli create stream stream1 '()WITH(FORMAT="JSON",TYPE="edgex")'
# rule 1 
edgexfoundry.kuiper-cli create rule rule_log '
{
   "sql":"SELECT * from stream1",
   "actions":[
      {
         "log":{
            
         }
      }
   ]
}' 
# rules 2
edgexfoundry.kuiper-cli create rule rule_edgex_message_bus '
{
   "sql":"SELECT * from stream1",
   "actions": [
      {
         "edgex": {
            "connectionSelector": "edgex.redisMsgBus",
            "topicPrefix": "edgex/events/device", 
            "messageType": "request",
            "deviceName": "device-test"
         }
      }
   ]
}'
```
Expect:
```
edgexfoundry.kuiper-cli getstatus rule rule_log
```
```
Connecting to 127.0.0.1:20498... 
{
  "source_stream1_0_records_in_total": 214983,
  "source_stream1_0_records_out_total": 214983,
  "source_stream1_0_exceptions_total": 0,
  "source_stream1_0_process_latency_us": 0,
  "source_stream1_0_buffer_length": 1,
  "source_stream1_0_last_invocation": "2022-03-17T15:09:30.757436",
  "op_2_project_0_records_in_total": 214982,
  "op_2_project_0_records_out_total": 214981,
  "op_2_project_0_exceptions_total": 0,
  "op_2_project_0_process_latency_us": 0,
  "op_2_project_0_buffer_length": 0,
  "op_2_project_0_last_invocation": "2022-03-17T15:09:30.757457",
  "sink_log_0_0_records_in_total": 214981,
  "sink_log_0_0_records_out_total": 214981,
  "sink_log_0_0_exceptions_total": 0,
  "sink_log_0_0_process_latency_us": 10,
  "sink_log_0_0_buffer_length": 0,
  "sink_log_0_0_last_invocation": "2022-03-17T15:09:30.757269"
}
```
```
edgexfoundry.kuiper-cli getstatus rule rule_edgex_message_bus
```
```
Connecting to 127.0.0.1:20498... 
{
  "source_stream1_0_records_in_total": 928309,
  "source_stream1_0_records_out_total": 928309,
  "source_stream1_0_exceptions_total": 0,
  "source_stream1_0_process_latency_us": 0,
  "source_stream1_0_buffer_length": 0,
  "source_stream1_0_last_invocation": "2022-03-17T15:10:17.534734",
  "op_2_project_0_records_in_total": 928309,
  "op_2_project_0_records_out_total": 928309,
  "op_2_project_0_exceptions_total": 0,
  "op_2_project_0_process_latency_us": 2,
  "op_2_project_0_buffer_length": 0,
  "op_2_project_0_last_invocation": "2022-03-17T15:10:17.534737",
  "sink_edgex_0_0_records_in_total": 928267,
  "sink_edgex_0_0_records_out_total": 928266,
  "sink_edgex_0_0_exceptions_total": 0,
  "sink_edgex_0_0_process_latency_us": 52,
  "sink_edgex_0_0_buffer_length": 40,
  "sink_edgex_0_0_last_invocation": "2022-03-17T15:10:17.534711"
}
```
```
edgexfoundry.kuiper-cli show rules
```
```
[
  {
    "id": "rule_edgex_message_bus",
    "status": "Running"
  },
  {
    "id": "rule_log",
    "status": "Running"
  }
]
```

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -476,10 +471,15 @@ apps:
...
     adapter: full
+    after:
+      - security-bootstrapper-redis
     command: bin/kuiperd
     daemon: simple
     environment:
       KuiperBaseKey: $SNAP_DATA/kuiper
+      VAULT_TOKEN_FILE: $SNAP_DATA/secrets/edgex-ekuiper/secrets-token.json
+      SOURCE_FILE: $SNAP_DATA/kuiper/etc/sources/edgex.yaml 
+      CONNECTIONS_FILE: $SNAP_DATA/kuiper/etc/connections/connection.yaml
```
Comment (farshidtz) [inline snap/snapcraft.yaml:482]: CONNECTIONS_FILE -> CONNECTION_FILE

**Review by farshidtz — CHANGES_REQUESTED** [review]

File: snap/local/hooks/cmd/install/install.go [hunk 2]
```diff
@@ -204,6 +207,24 @@ func installKuiper() error {
...
+		Args:   []string{copyScriptPath},
+		Stdout: os.Stdout,
+		Stderr: os.Stdout,
+	}
+
+	err = cmdCopyKuiper.Run()
+	if err != nil {
+		return err
+	}
+
```
Comment (farshidtz) [inline snap/local/hooks/cmd/install/install.go:227]: The edge-ekuiper snap sets `KuiperBaseKey: $SNAP_DATA` while this snap has it set to `KuiperBaseKey: $SNAP_DATA/kuiper`. As a result, an additional script is called here to copy them (why not move) to copy ./etc from former to latter directory.

Can't we just make the [install hook](https://github.com/canonical/edgex-ekuiper-snap/blob/main/snap/hooks/install) of edgex-ekuiper take an environment variable as input base path which is set to `$SNAP_DATA` by default and set to `$SNAP_DATA/kuiper` by passing [Cmd.Env](https://pkg.go.dev/os/exec#Cmd) when called here?

**Review by farshidtz — CHANGES_REQUESTED** [review]
Thanks @MonicaisHer. 
It looks good. Added a few minor comments.

File: snap/local/hooks/cmd/install/install.go [hunk 3]
```diff
@@ -187,22 +187,23 @@ func installDevProfiles() error {
...
+		Path: filePath,
+		Env:  append(os.Environ(), "KUIPER_BASE_KEY="+hooks.SnapData+"/kuiper"),
 	}
 
-	err = cmdSetupKuiper.Run()
+	stdoutStderr, err := cmdSetupKuiper.CombinedOutput()
 	if err != nil {
 		return err
 	}
+	fmt.Println(fmt.Sprintf("edgexfoundry:install-kuiper: %s", stdoutStderr))
```
[the remark this answers is not in the document]
Comment (MonicaisHer) [reply snap/local/hooks/cmd/install/install.go:206]: error is logged [here](https://github.com/edgexfoundry/edgex-go/blob/main/snap/local/hooks/cmd/install/install.go#L395)

[the remark this answers is not in the document]
Comment (farshidtz) [reply snap/local/hooks/cmd/install/install.go:206]: I meant that standard errors aren't logged. As far as I know, `err` doesn't include the standard errors.

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -865,36 +867,10 @@ parts:
...
-        -e 's@ip\: 0.0.0.0@ip\: localhost@' \
-        -e 's@restIp\: 0.0.0.0@restIp\: localhost@' \
-        -e 's@restPort\: 9081@restPort\: 59720@' \
-        $SNAPCRAFT_PART_INSTALL/etc/kuiper.yaml
-
-      install -DT "$SNAPCRAFT_PART_SRC/LICENSE" \
-         "$SNAPCRAFT_PART_INSTALL/usr/share/doc/kuiper/LICENSE"
-    stage:
-      - -etc/mqtt_source.yaml
-      - -etc/services/
```
[the remark this answers is not in the document]
Comment (MonicaisHer) [reply snap/snapcraft.yaml:898]: Do you mean to include files using the `-` prefix? `+` prefix doesn't work with `stage:`

[the remark this answers is not in the document]
Comment (farshidtz) [reply snap/snapcraft.yaml:898]: The `-` prefix is to exclude. No prefix means inclusion (not `+`). Not to be confused with the hyphen used in YAML list. See https://snapcraft.io/docs/snapcraft-filesets which has the same syntax.

Anyway, let's leave it for future work.

**Review by farshidtz — APPROVED** [review]

**Pull request merged**